File: prolog/tokenizer.pl
:- module(tokenizer, [tokenize/2]).

tokenize(StringCodes, Tokens) :- phrase(tok(Tokens), StringCodes).

tok([Token | Ts]) --> ch(punct, [C|Cs]), {lex([C|Cs], Token)}, !, tok(Ts).
tok([Token | Ts]) --> ch(alpha, [C|Cs]), {lex([C|Cs], Token)}, !, tok(Ts).
tok([int(Number) | Ts]) --> ch(digit, [C|Cs]), {number_codes(Number, [C|Cs])}, !, tok(Ts).
tok([ident(Atom) | Ts]) --> ch(alpha, [C|Cs]), {atom_codes(Atom, [C|Cs])}, !, tok(Ts).
tok(Ts) --> ch(space, [_]), !, tok(Ts).
tok([eof]) --> [].

ch(Type, [C|Cs]) --> [C], {code_type(C, Type)}, ch(Type, Cs).
ch(_, []) --> [].

lex([61], equal).
lex([43], plus).
lex([44], comma).
lex([59], semicolon).
lex([40], lparen).
lex([41], rparen).
lex([123], lsquirly).
lex([125], rsquirly).
lex([33], bang).
lex([45], dash).
lex([47], fslash).
lex([42], asterisk).
lex([60], lt).
lex([62], gt).
lex([61, 61], eq).
lex([33, 61], neq).
lex([62, 61], geq).
lex([60, 61], leq).
lex([108, 101, 116], let).
lex([102, 110], function).
lex([105, 102], if).
lex([101, 108, 115, 101], else).
lex([116, 114, 117, 101], true).
lex([102, 97, 108, 115, 101], false).
lex([114, 101, 116, 117, 114, 110], return).

File: ruby/test/tokenizer.rb
require 'rspec'
require_relative '../src/lexer'

describe Tokenizer do
  it 'Test Tokenizer' do
    input = %q(
        let five = 5;
        let ten = 10;

        let add = fn(x, y) {
            x + y;
        };

        let result = add(five, ten);
        !-/*5;
        5 < 10 > 5;

        if (5 < 10) {
            return true;
        } else {
            return false;
        }

        10 == 10;
        10 != 9;
    )
    lexer = Tokenizer.new(input)

    tokens = [
        { type: TokenType::Let, literal: 'let' },
        { type: TokenType::Ident, literal: 'five' },
        { type: TokenType::Equal, literal: '=' },
        { type: TokenType::Int, literal: '5' },
        { type: TokenType::Semicolon, literal: ';' },
        { type: TokenType::Let, literal: 'let' },
        { type: TokenType::Ident, literal: 'ten' },
        { type: TokenType::Equal, literal: '=' },
        { type: TokenType::Int, literal: '10' },
        { type: TokenType::Semicolon, literal: ';' },
        { type: TokenType::Let, literal: 'let' },
        { type: TokenType::Ident, literal: 'add' },
        { type: TokenType::Equal, literal: '=' },
        { type: TokenType::Function, literal: 'fn' },
        { type: TokenType::LParen, literal: '(' },
        { type: TokenType::Ident, literal: 'x' },
        { type: TokenType::Comma, literal: ',' },
        { type: TokenType::Ident, literal: 'y' },
        { type: TokenType::RParen, literal: ')' },
        { type: TokenType::LSquirly, literal: '{' },
        { type: TokenType::Ident, literal: 'x' },
        { type: TokenType::Plus, literal: '+' },
        { type: TokenType::Ident, literal: 'y' },
        { type: TokenType::Semicolon, literal: ';' },
        { type: TokenType::RSquirly, literal: '}' },
        { type: TokenType::Semicolon, literal: ';' },
        { type: TokenType::Let, literal: 'let' },
        { type: TokenType::Ident, literal: 'result' },
        { type: TokenType::Equal, literal: '=' },
        { type: TokenType::Ident, literal: 'add' },
        { type: TokenType::LParen, literal: '(' },
        { type: TokenType::Ident, literal: 'five' },
        { type: TokenType::Comma, literal: ',' },
        { type: TokenType::Ident, literal: 'ten' },
        { type: TokenType::RParen, literal: ')' },
        { type: TokenType::Semicolon, literal: ';' },
        { type: TokenType::Bang, literal: '!' },
        { type: TokenType::Minus, literal: '-' },
        { type: TokenType::Slash, literal: '/' },
        { type: TokenType::Asterisk, literal: '*' },
        { type: TokenType::Int, literal: '5' },
        { type: TokenType::Semicolon, literal: ';' },
        { type: TokenType::Int, literal: '5' },
        { type: TokenType::LT, literal: '<' },
        { type: TokenType::Int, literal: '10' },
        { type: TokenType::GT, literal: '>' },
        { type: TokenType::Int, literal: '5' },
        { type: TokenType::Semicolon, literal: ';' },
        { type: TokenType::If, literal: 'if' },
        { type: TokenType::LParen, literal: '(' },
        { type: TokenType::Int, literal: '5' },
        { type: TokenType::LT, literal: '<' },
        { type: TokenType::Int, literal: '10' },
        { type: TokenType::RParen, literal: ')' },
        { type: TokenType::LSquirly, literal: '{' },
        { type: TokenType::Return, literal: 'return' },
        { type: TokenType::True, literal: 'true' },
        { type: TokenType::Semicolon, literal: ';' },
        { type: TokenType::RSquirly, literal: '}' },
        { type: TokenType::Else, literal: 'else' },
        { type: TokenType::LSquirly, literal: '{' },
        { type: TokenType::Return, literal: 'return' },
        { type: TokenType::False, literal: 'false' },
        { type: TokenType::Semicolon, literal: ';' },
        { type: TokenType::RSquirly, literal: '}' },
        { type: TokenType::Int, literal: '10' },
        { type: TokenType::EQ, literal: '==' },
        { type: TokenType::Int, literal: '10' },
        { type: TokenType::Semicolon, literal: ';' },
        { type: TokenType::Int, literal: '10' },
        { type: TokenType::NotEQ, literal: '!=' },
        { type: TokenType::Int, literal: '9' },
    ]

    tokens.each do |token|
        expect(lexer.next_token).to eq(token)
    end
   end
end